### PR TITLE
[Internal] TransactionalBatchItemRequestOptions: Adds WriteRequestProperties and GetRequestPropertiesSerializationLength

### PR DESCRIPTION
**Pull request by vivekr20**
This change moves the logic of writing request properties in ItemBatchOperation.WriteOperation into an internal virtual method on TransactionalBatchItemRequestOptions. By doing this, different internal implementations of TransactionalBatchItemRequestOptions can override this method to allow writing properties differently.

## Type of change
- [] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs [hunk 1]
```diff
@@ -304,6 +312,10 @@ internal int GetApproximateSerializedLength()
                         {
                             length += binaryId.Length;
                         }
+                        else if (binaryIdObj is ReadOnlyMemory<byte> binaryIdRom)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs:315]: As discussed why not move this to the the batch request option as a internal virtual method. Then Friends/Compute can override it have whatever logic neccesary.

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs [hunk 2]
```diff
@@ -189,57 +188,10 @@ internal static Result WriteOperation(ref RowWriter writer, TypeArgument typeArg
                     }
                 }
 
-                if (options.Properties != null)
+                r = operation.RequestOptions.WriteRequestProperties(ref writer, pkWritten);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs:191]: Couple lines above there is code that simplifies the reference to `options`
```suggestion
                r = options.WriteRequestProperties(ref writer, pkWritten);
```

File: Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchItemRequestOptions.cs [hunk 3]
```diff
@@ -48,5 +51,63 @@ internal static TransactionalBatchItemRequestOptions FromItemRequestOptions(Item
             };
             return batchItemRequestOptions;
         }
+
+        internal virtual Result WriteRequestProperties(ref RowWriter writer, bool pkWritten)
+        {
+            if (this.Properties != null)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchItemRequestOptions.cs:57]: nit: reduce nested if
```c#
if (this.Properties == null)
{
   return Result.Success;
}
```

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/ItemBatchOperationTests.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    using Moq;
+
+    /// <summary>
+    /// Tests for <see cref="ItemBatchOperation"/>.
+    /// </summary>
+    [TestClass]
+    public class ItemBatchOperationTests
+    {
+        [TestMethod]
+        public void TestWriteOperationWithBinaryIdByteArray()
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/ItemBatchOperationTests.cs:23]: nit: Would this test be better located in https://github.com/Azure/azure-cosmos-dotnet-v3/blob/af9b009971a1bfc83d5ea9eeba53445b8c496fc6/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/BatchSchemaTests.cs#L24 ?

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs (hunk 2 above)
Comment (vivekr20) [reply Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs:191]: ack, fixing this in the next iteration.

File: Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs [hunk 5]
```diff
@@ -304,6 +256,10 @@ internal int GetApproximateSerializedLength()
                         {
                             length += binaryId.Length;
                         }
+                        else if (binaryIdObj is ReadOnlyMemory<byte> binaryIdRom)
+                        {
+                            length += binaryIdRom.Length;
```
[the remark this answers is not in the document]
Comment (vivekr20) [reply Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs:261]: good catch, let me add another method for this.

[the remark this answers is not in the document]
Comment (vivekr20) [reply Microsoft.Azure.Cosmos/src/Batch/ItemBatchOperation.cs]: addressed in recent PR commit.

**Review by j82w — APPROVED** [review]

**Review by ealsur — APPROVED** [review]

**Pull request merged**